File: src/android/app/src/main/java/com/crickettechnology/trees/QtNode.java
package com.crickettechnology.trees;

import java.util.*;

// Quadtree node
class QtNode<T extends QtElement>
{
    public QtNode(QtRect rect)
    {
        this.rect = rect;
    }

    // find element nearest to the given location
    public T findNearest(final QtLocation location)
    {
        if (this.rect.contains(location))
        {
            return null;
        }

        ArrayList<T> elems = new ArrayList<>();
        if (this.nodes == null)
        {
            elems.addAll(this.elements);
        }
        else
        {
            for (QtNode<T> node: this.nodes)
            {
                T nearest = node.findNearest(location);
                if (node != null)
                {
                    elems.add(nearest);
                }
            }
        }

        return Collections.min(elems, 
            new Comparator<T>()
            {
                @Override
                public int compare(T lhs, T rhs) 
                {
                    double lhsDist = lhs.getLocation().distanceSquared(location);
                    double rhsDist = rhs.getLocation().distanceSquared(location);
                    return Double.compare(lhsDist, rhsDist);
                }
            });
    }

    // find all elements contained in the given rectangle
    public List<T> findContained(QtRect rect)
    {
        ArrayList<T> results = new ArrayList<>();
        if (this.nodes == null)
        {
            for (T elem: this.elements)
            {
                if (rect.contains(elem.getLocation()))
                {
                    results.add(elem);
                }
            }
        }
        else
        {
            for (QtNode<T> node: this.nodes)
            {
                if (node.rect.intersects(rect))
                {
                    results.addAll(node.findContained(rect));
                }
            }
        }
        return results;
    }

    // find all rectangles intersected by the given rectangle
    public List<QtRect> findContainedRects(QtRect rect)
    {
        ArrayList<QtRect> results = new ArrayList<>();
        if (this.rect.intersects(rect) && !this.rect.contains(rect))
        {
            results.add(this.rect);
        }
        if (this.nodes != null)
        {
            for (QtNode<T> node: this.nodes)
            {
                results.addAll(node.findContainedRects(rect));
            }
        }
        return results;
    }

    public void add(T element)
    {
        if (this.nodes == null)
        {
            this.elements.add(element);
            if (this.elements.size() > k_maxElements)
            {
                subdivide();
            }
        }
        else
        {
            place(element);
        }
    }

    public int countElements()
    {
        if (this.nodes == null)
        {
            return this.elements.size();
        }
        else
        {
            int count = 0;
            for (QtNode<T> node: this.nodes)
            {
                count += node.countElements();
            }
            return count;
        }
    }

    public int countNodes()
    {
        if (this.nodes == null)
        {
            return 1;
        }
        else
        {
            int count = 1;
            for (QtNode<T> node: this.nodes)
            {
                count += node.countNodes();
            }
            return count;
        }
    }

    public int countLevels()
    {
        if (this.nodes == null)
        {
            return 1;
        }
        else
        {
            int count = 0;
            for (QtNode<T> node: this.nodes)
            {
                count = Math.max(count, node.countElements());
            }
            return 1 + count;
        }
    }

    public QtNode<T> clone()
    {
        QtNode<T> node = new QtNode<>(this.rect);
        if (this.nodes == null)
        {
            node.elements.addAll(this.elements);
        }
        else
        {
            node.nodes = new QtNode[]
            {
                node.nodes[0].clone(),
                node.nodes[1].clone(),
                node.nodes[2].clone(),
                node.nodes[3].clone(),
            };
        }
        return node;
    }

    ////////////////////////////////////////

    private QtRect rect;
    private QtNode<T>[] nodes;
    private ArrayList<T> elements = new ArrayList<>();
    private static final int k_maxElements = 10;

    private void subdivide()
    {
        assert this.nodes == null;
        this.nodes = new QtNode[]
        {
            new QtNode(this.rect.getQuarter(0,0)),
            new QtNode(this.rect.getQuarter(0,1)),
            new QtNode(this.rect.getQuarter(1,0)),
            new QtNode(this.rect.getQuarter(1,1))
        };
        for (T elem: this.elements)
        {
            place(elem);
        }
        this.elements.clear();
    }

    private void place(T elem)
    {
        assert(this.nodes.length == 4);
        for (QtNode<T> node: this.nodes)
        {
            if (node.rect.contains(elem.getLocation()))
            {
                node.add(elem);
                break;
            }
        }
    }
}

File: src/ios/trees/QtNode.swift
import Foundation

// Quadtree node
class QtNode<T: QtElement>
{
    init(rect: QtRect)
    {
        self.rect = rect
    }

    // find element nearest to the given location
    func findNearest(location: QtLocation, distanceScale: QtDistanceScale? = nil) -> T?
    {
        if (!self.rect.contains(location:location))
        {
            return nil
        }

        let distanceScale = distanceScale ?? QtDistanceScale(location: location)

        var elems = [T]()
        if (self.nodes.isEmpty)
        {
            elems.append(contentsOf:self.elements)
        }
        else
        {
            for node in self.nodes
            {
                if let nearest = node.findNearest(location:location, distanceScale: distanceScale)
                {
                    elems.append(nearest)
                }
            }
        }

        return elems.min(by: 
                { 
                    x, y in
                    let xLoc = x.location
                    let yLoc = y.location
                    return distanceScale.distanceSquared(xLoc, location) < distanceScale.distanceSquared(yLoc, location)
                })
    }

    // find all elements contained in the given rectangle
    func findContained(rect: QtRect) -> [T]
    {
        var results = [T]()
        if self.nodes.isEmpty
        {
            for elem in self.elements
            {
                if rect.contains(location:elem.location)
                {
                    results.append(elem)
                }
            }
        }
        else
        {
            for node in self.nodes
            {
                if node.rect.intersects(rect:rect)
                {
                    results.append(contentsOf: node.findContained(rect:rect))
                }
            }
        }
        return results
    }

    // find all rectangles intersected by the given rectangle
    func findContainedRects(rect: QtRect) -> [QtRect]
    {
        var results = [QtRect]()
        if self.rect.intersects(rect: rect) && !self.rect.contains(rect:rect)
        {
            results.append(self.rect)
        }
        for node in self.nodes
        {
            if node.rect.intersects(rect: rect)
            {
                results.append(contentsOf: node.findContainedRects(rect:rect))
            }
        }
        return results
    }

    func add(_ element: T) 
    {
        if (self.nodes.isEmpty)
        {
            self.elements.append(element)
            if (self.elements.count > k_maxElements)
            {
                subdivide()
            }
        }
        else
        {
            place(element)
        }
    }

    func countElements() -> Int
    {
        if (self.nodes.isEmpty)
        {
            return self.elements.count
        }
        else
        {
            return self.nodes.reduce(0, { sum, node in sum + node.countElements() })
        }
    }

    func countNodes() -> Int
    {
        if (self.nodes.isEmpty)
        {
            return 1
        }
        else
        {
            return self.nodes.reduce(1, { sum, node in sum + node.countNodes() })
        }
    }

    func countLevels() -> Int
    {
        if (self.nodes.isEmpty)
        {
            return 1
        }
        else
        {
            return 1 + self.nodes.map({ $0.countLevels() }).max()!
        }
    }

    func clone() -> QtNode<T>
    {
        let node = QtNode(rect: self.rect)
        if self.nodes.isEmpty
        {
            node.elements.append(contentsOf:self.elements)
        }
        else
        {
            for n in self.nodes
            {
                node.nodes.append(n.clone())
            }
        }
        return node
    }

    /*
    func print(prefix: String = "")
    {
        if (self.nodes.isEmpty)
        {
            Swift.print("\(prefix)\(self.elements.count)")
        }
        else
        {
            for node in self.nodes
            {
                node.print(prefix: prefix + "-")
            }
        }
    }
    */

    ////////////////////////////////////////

    private var rect: QtRect
    private var nodes = [QtNode]()
    private var elements = [T]()
    private let k_maxElements = 10

    private func subdivide()
    {
        assert(self.nodes.isEmpty)
        self.nodes = 
        [
            QtNode(rect: self.rect.getQuarter(0,0)),
            QtNode(rect: self.rect.getQuarter(0,1)),
            QtNode(rect: self.rect.getQuarter(1,0)),
            QtNode(rect: self.rect.getQuarter(1,1))
        ]
        for elem in self.elements
        {
            place(elem)
        }
        self.elements.removeAll()
    }

    private func place(_ elem: T)
    {
        assert(self.nodes.count == 4)
        for node in self.nodes
        {
            if (node.rect.contains(location:elem.location))
            {
                node.add(elem)
                break
            }
        }
    }
}
